src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1169 @@
+
+        pub fn deinit(self: ExecuteDetailsTimings, allocator: std.mem.Allocator) void {
+            var per_program_timings = self.per_program_timings;
+            for (self.per_program_timings.values()) |pt| pt.deinit(allocator);
+            per_program_timings.deinit(allocator);
+        }
+
+        pub fn clone(
+            self: ExecuteDetailsTimings,
+            allocator: std.mem.Allocator,
+        ) std.mem.Allocator.Error!ExecuteDetailsTimings {
+            var ppt: PerProgramTimings = .{};
+            errdefer ppt.deinit(allocator);
+            errdefer for (ppt.values()) |program_timing| program_timing.deinit(allocator);
+            try ppt.ensureTotalCapacity(allocator, self.per_program_timings.count());
+            for (
+                self.per_program_timings.keys(),
+                self.per_program_timings.values(),
+            ) |k, v| ppt.putAssumeCapacityNoClobber(k, try v.clone(allocator));
+
+            var copy = self;
+            copy.per_program_timings = ppt;
+            return copy;
+        }
+
+        pub fn eql(self: *const ExecuteDetailsTimings, other: *const ExecuteDetailsTimings) bool {
+            inline for (@typeInfo(ExecuteDetailsTimings).Struct.fields) |field| {
+                const self_field = &@field(self, field.name);
+                const other_field = &@field(other, field.name);
+                switch (field.type) {
+                    Saturating(u64),
+                    => if (self_field.toInt() != other_field.toInt()) {
+                        return false;
+                    },
+                    std.AutoArrayHashMapUnmanaged(Pubkey, ProgramTiming),
+                    => {
+                        if (self_field.count() != other_field.count()) {
+                            return false;
+                        }
+                        for (self_field.keys(), self_field.values()) |self_k, self_v| {
+                            const other_entry = other_field.getEntry(self_k) orelse return false;
+                            if (!self_k.equals(other_entry.key_ptr)) return false;
+                            if (!self_v.eql(other_entry.value_ptr)) return false;
+                        }
+                    },
+                    else => comptime unreachable,
+                }
+            }
+            return true;
+        }
+    };
+
+    pub const ExecuteProcessInstructionTimings = struct {
+        total_us: Saturating(u64),
+        verify_caller_us: Saturating(u64),
+        process_executable_chain_us: Saturating(u64),
+        verify_callee_us: Saturating(u64),
+    };
+
+    pub const ExecuteAccessoryTimings = struct {
+        feature_set_clone_us: Saturating(u64),
+        get_executors_us: Saturating(u64),
+        process_message_us: Saturating(u64),
+        process_instructions: ExecuteProcessInstructionTimings,
+    };
+};
+
+test "ProgressMap memory ownership" {
```

Comment:
It would be nice to also have the same test coverage as agave's 4 tests in progress_map.rs, plus supporting whatever functions are used in those tests.

Reply:
00f61eba5189cd56a229187677ee292f8dbe0545
Implemented three of the four (3/4) tests, I'll implement the fourth in the follow up PR, since it tests some other unimplemented bits.